### PR TITLE
misc: use assert_never to enforce exhaustive matches

### PR DESCRIPTION
**Pull request by superlopuh**
Doing some cleanup

### CODE REVIEW

File: xdsl/dialects/onnx.py [hunk 1]
```diff
@@ -276,13 +270,10 @@ def __init__(
...
-            or not isinstance(tensor_c_type := self.tensor_c.type, TensorType)
-            or not isinstance(res_tensor_type := self.res_tensor.type, TensorType)
-        ):
-            assert False, "onnx elementwise operation operands and result must be of type TensorType"
+        assert isinstance(tensor_a_type := self.tensor_a.type, TensorType)
+        assert isinstance(tensor_b_type := self.tensor_b.type, TensorType)
+        assert isinstance(tensor_c_type := self.tensor_c.type, TensorType)
+        assert isinstance(res_tensor_type := self.res_tensor.type, TensorType)
 
         tensor_a_type = cast(TensorType[Attribute], tensor_a_type)
```
[the remark this answers is not in the document]
Comment (alexarice) [reply xdsl/dialects/onnx.py:278]: Same in other places

File: xdsl/transforms/convert_ptr_to_riscv.py [hunk 2]
```diff
@@ -91,23 +90,19 @@ def match_and_rewrite(self, op: ptr.LoadOp, rewriter: PatternRewriter, /):
...
+            float_type = cast(AnyFloat, op.res.type)
+            match float_type:
+                case Float32Type():
+                    lw_op = riscv.FLwOp(addr, 0, comment="load float from pointer")
+                case Float64Type():
+                    lw_op = riscv.FLdOp(addr, 0, comment="load double from pointer")
+                case _:
+                    raise DiagnosticException(
+                        f"Unexpected floating point type {float_type}"
+                    )
```
Comment (alexarice) [inline xdsl/transforms/convert_ptr_to_riscv.py:105]: Can the register definitely not take a different type?

File: xdsl/transforms/lower_affine.py [hunk 3]
```diff
@@ -29,34 +29,35 @@ def affine_expr_ops(
...
-            case AffineBinaryOpKind.FloorDiv:
-                op = arith.FloorDivSI(lhs_val, rhs_val)
-            case AffineBinaryOpKind.CeilDiv:
-                op = arith.CeilDivSI(lhs_val, rhs_val)
-
-        return [*lhs_ops, *rhs_ops, op], op.result
-
-    assert False, "Unreachable"
+    match expr:
+        case AffineConstantExpr():
```
Comment (alexarice) [inline xdsl/transforms/lower_affine.py:33]: Is there a check that this actually works? Seems surprising to me that this does an isinstance check

**Review by alexarice — COMMENTED** [review]
Seems good, just left a few comments. I'm also not sure what the advantage of `assert_never` over `assert False` is.

File: xdsl/dialects/onnx.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/dialects/onnx.py:278]: unfortunately yes, I tried to remove them, but sometimes we actually get the scalar type of the tensor type, and that's a full-on Unknown value, as opposed to an Unknown argument, so can't be pyright-ignored away.

File: xdsl/transforms/convert_ptr_to_riscv.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/transforms/convert_ptr_to_riscv.py:105]: Hmm, this should probably have a different message, as it's more of a not implemented situation, I'll add this.

File: xdsl/transforms/lower_affine.py (hunk 3 above)
Comment (superlopuh) [reply xdsl/transforms/lower_affine.py:33]: yep plenty of checks, this is the weird syntax python uses for isinstance checks in match statements

**Review by alexarice — APPROVED** [review]
I'm still not sure what the advantage of this is but seems ok

**Comment by superlopuh** [comment]
`assert_never` is a Pyright error if the type of the values passed in is not `Never`, so we get type system support that's a bit more explicit. It raises an error at runtime if called, like `assert False`, and is just a bit more explicit about the intent.

**Pull request merged**